Reworked cortex m systick

avolkov-1221 opened the pull request:
Hello,

This commit added missing support of the EXTERNAL clock to the Cortex-M Systick driver.
Also some minor refactoring has been done to fix following problems: old IRQ API usage and non-atomic writing to the control registers.

Regards
Andrey VOLKOV

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/56117

andyross commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -396,6 +396,7 @@ mstar	MStar Semiconductor, Inc. (acquired by MediaTek Inc.)
 mti	Imagination Technologies Ltd. (formerly MIPS Technologies Inc.)
 multi-inno	Multi-Inno Technology Co.,Ltd
 mundoreader	Mundo Reader S.L.
+munic	MUNIC SA
```
I don't see any DTS in this PR that would use this?  The hardware IP in question is from ARM, obviously.

andyross commented on drivers/timer/Kconfig.cortex_m_systick:
```diff
@@ -24,3 +24,15 @@ config CORTEX_M_SYSTICK_INSTALL_ISR
+
+config CORTEX_M_SYSTICK_EXT_FRQ
+	int "External clock frequency"
+	depends on CORTEX_M_SYSTICK_EXT_CLOCK
```
Traditionally timer drivers have been only loosely coupled to devicetree (it's just history, they're very low level and they predate DTS), but as we add new configurables it's probably best to do it where it belongs.

andyross reviewed (changes requested):
Some quick notes, not a full review.  Leaving a -1 to note that we really need to clarify exactly what the logic changes are and how they work, ideally via a series of small changes interspersed with refactoring patches that make clear improvements (or at worst just do things like reorder functions or move blocks of code).

@npitre probably wants to review too.

avolkov-1221 commented:
Hello Andy,

Le 14/03/2023 à 16:26, Andy Ross a écrit :
>
> ***@***.**** requested changes on this pull request.
>
> Some quick notes, not a full review. Leaving a -1 to note that we 
> really need to clarify exactly what the logic changes are and how they 
> work, ideally via a series of small changes interspersed with 
> refactoring patches that make clear improvements (or at worst just do 
> things like reorder functions or move blocks of code).
>
> @npitre <https://github.com/npitre> probably wants to review too.
>
Systick of Cortex-M can use two clock sources as a reference frequency: 
the CPU's and the so-called "external", defined by chip manufacturer. At 
least Microchip(Atmel), STM and Renesas have implemented this functionality.

However the current driver lacks support for an "external" clock source 
at all, and to make the fact worse, the "external" clock has nothing 
common with the CPU's clock, so the whole driver's time calculation 
logic has been completely changed.Also small roughnessin the form of 
"read-modify-write" instead of atomic writing are not so well looking.As 
hencebreaking the patch into really small pieces will not be so easy 
(driver itself is not so huge), but I will try to do it, at least 
rearranging code blocks.

> ------------------------------------------------------------------------
>
> In dts/bindings/vendor-prefixes.txt 
> <https://github.com/zephyrproject-rtos/zephyr/pull/55762#discussion_r1135728743>:
>
> > @@ -396,6 +396,7 @@ mstar	MStar Semiconductor, Inc. (acquired by MediaTek Inc.)
>   mti	Imagination Technologies Ltd. (formerly MIPS Technologies Inc.)
>   multi-inno	Multi-Inno Technology Co.,Ltd
>   mundoreader	Mundo Reader S.L.
> +munic	MUNIC SA
>
> I don't see any DTS in this PR that would use this? The hardware IP in 
> question is from ARM, obviously.
>
Oh yes, sorry, it was the top of my patch series, so I've rearranged it 
and will send it later.

> ------------------------------------------------------------------------
>
> In drivers/timer/Kconfig.cortex_m_systick 
> <https://github.com/zephyrproject-rtos/zephyr/pull/55762#discussion_r1135730372>:
>
> > @@ -24,3 +24,15 @@ config CORTEX_M_SYSTICK_INSTALL_ISR
>   	help
>   	  This option should be selected by SysTick-based drivers so that the
>   	  sys_clock_isr() function is installed.
> +
> +if CORTEX_M_SYSTICK
> +config CORTEX_M_SYSTICK_EXT_CLOCK
> +	bool "Use external reference clock as SysTick source instead of CPU one"
> +	default n
> +	depends on CORTEX_M_SYSTICK
> +
> +config CORTEX_M_SYSTICK_EXT_FRQ
> +	int "External clock frequency"
> +	depends on CORTEX_M_SYSTICK_EXT_CLOCK
>
> Traditionally timer drivers have been only loosely coupled to 
> devicetree (it's just history, they're very low level and they predate 
> DTS), but as we add new configurables it's probably best to do it 
> where it belongs.
>
Ok, noted, I'm a newbie with the Zephyr, so will know.

> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/zephyrproject-rtos/zephyr/pull/55762#pullrequestreview-1339562429>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/A6DRPADVDU742JTHNVSRGBDW4CE25ANCNFSM6AAAAAAV2MSRUU>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

andyross reviewed (changes requested):
OK, this is easier, thanks.  Some notes.  The -1 is for the merge commit, we don't allow those in PRs.  The rest are mostly nitpicks, but a few probably bear some attention.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -115,9 +115,8 @@ static uint32_t elapsed(void)
 
 /* Callout out of platform assembly, not hooked via IRQ_CONNECT... */
-void sys_clock_isr(void *arg)
+ISR_DIRECT_DECLARE(sys_clock_isr)
```
I don't know when indirect->direct changing was really happened, but actual "sys_clock_isr" is a direct isr call. And yes, I was sure that ISR_DIRECT_DECLARE does this: "They must be implemented in such a way that it is safe to put them directly in the vector table.  For ISRs written in C, The ISR_DIRECT_DECLARE() macro will do this automatically.", but actually it does not. Will fix my implementation.

BTW IMHO the current driver should also be marked "broken" because it still uses the normal C calling convention.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -25,86 +27,89 @@
-				SYSTICK_CSR_TICKINT	| \
+#if USE_EXT_CLOCK
+#define SYS_CLOCK_FRQ	DT_PROP(DT_NODELABEL(systick), external_frequency)
+#define cyc_per_tick	(SYS_CLOCK_FRQ / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
```
Ok, accepted.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -11,8 +12,9 @@
 
-#define COUNTER_MAX 0x00ffffff
-#define TIMER_STOPPED 0xff000000
+#define TICKLESS (IS_ENABLED(CONFIG_TICKLESS_KERNEL))
```
I agree, this came from old code and has been kept for easier merging.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -25,86 +27,89 @@
-#define MAX_TICKS ((k_ticks_t)(COUNTER_MAX / CYC_PER_TICK) - 1)
-#define MAX_CYCLES (MAX_TICKS * CYC_PER_TICK)
+/* max number of ticks we can load into the timer in one shot */
+#define MAX_TICKS (COUNTER_MAX * SYS_CLOCK_FRQ / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
```
Yep.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -121,184 +126,198 @@ static uint32_t elapsed(void)
-		dticks = (cycle_count - announced_cycles) / CYC_PER_TICK;
-		announced_cycles += dticks * CYC_PER_TICK;
-		sys_clock_announce(dticks);
+		dticks = clk / cyc_per_tick;
```
Ok will revert it. 

Just small clarification: "now" for me is "the current value of the timer's counter", but yes, the timer itself is restarted by set_timeout(), so the value can also be considered as "expired" too.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -121,184 +126,198 @@ static uint32_t elapsed(void)
-	sys_write32(last_load, SYSTICK_RVR);
-	sys_write32(0, SYSTICK_CVR); /* resets timer to last_load */
-	sys_write32(SYSTICK_CSR_INIT, SYSTICK_CSR);
+	irq_enable(SysTick_IRQn);
```
OMG, ok, fixed.

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -25,86 +27,89 @@
+#define cyc_per_tick	(SYS_CLOCK_FRQ / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define SYSTICK_CSR_INIT (SYSTICK_CSR_ENABLE | SYSTICK_CSR_TICKINT)
+#else
+#if  IS_ENABLED(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
```
Good news for me: CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME is a real headache and I was sure it's part of the API so I kept it.

avolkov-1221 commented:
> You moved the result of `CYC_PER_TICK` to a variable. When using an external clock, that variable is a `static const` which is just as good as a literal constant value.
> 
> Otherwise `cyc_per_tick` is not a const variable and you end up computing divisions with it at run time all the time. This should be done only if `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` is set. When not set, `sys_clock_hw_cycles_per_sec()` resolves to a constant at compile time and you want the compiler to possibly optimize those divisions by a constant when possible.

Ok, I've replaced the global var by inline 'get_cyc_per_tick'. I hope it closes the problem once and for all. It will also make things easier if a dynamic frequency management is ever added.

npitre commented:
I don't understand why you do those cycle conversions in
`sys_clock_cycle_get_32()` and `sys_clock_cycle_get_64()`.

If the external clock is a fixed value known at compile time then
`CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` should be the same as `SYS_CLOCK_FRQ`.
In other words, there shouldn't be any difference between return value of
`sys_clock_hw_cycles_per_sec()` and `SYS_CLOCK_FRQ`.

It is possible for Kconfig symbols to be initialized with a device tree 
entry value so no need to fear those getting out of sync.

npitre commented:
Quoting the documentation:

"
The kernel presents a "cycle" count via the `k_cycle_get_32` and 
`k_cycle_get_64` APIs. The intent is that this counter represents the 
fastest cycle counter that the operating system is able to present to the 
user (for example, a CPU cycle counter) and that the read operation is 
very fast.  The expectation is that very sensitive application code might 
use this in a polling manner to achieve maximal precision. The frequency 
of this counter is required to be steady over time, and is available from 
`sys_clock_hw_cycles_per_sec` (which on almost all platforms is a runtime 
constant that evaluates to `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`).
"

There is no presumption of "CPU's clock" here. It may well be equivalent to
the CPU's clock on some platforms but for all the boards I've worked with 
so far it isn't.

Furthermore, by performing a large division everytime those functions are 
called, they won't be "very fast" as the documentation implies it should.

And none of the other timer drivers perform such conversion either.

avolkov-1221 commented:
> Quoting the documentation:
> 
> " The kernel presents a "cycle" count via the `k_cycle_get_32` and `k_cycle_get_64` APIs. The intent is that this counter represents the fastest cycle counter that the operating system is able to present to the user (for example, a CPU cycle counter) and that the read operation is very fast. The expectation is that very sensitive application code might use this in a polling manner to achieve maximal precision. The frequency of this counter is required to be steady over time, and is available from `sys_clock_hw_cycles_per_sec` (which on almost all platforms is a runtime constant that evaluates to `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`). "
> 
> There is no presumption of "CPU's clock" here. It may well be equivalent to the CPU's clock on some platforms but for all the boards I've worked with so far it isn't.
> 
> Furthermore, by performing a large division everytime those functions are called, they won't be "very fast" as the documentation implies it should.
> 
> And none of the other timer drivers perform such conversion either.

I double-checked where this calculation came from and found that it refers to the epoch 2.7. So I removed the useless division.

avolkov-1221 committed:
cortex_m_systick: use ISR_DIRECT_DECLARE() instead of incorrect C-call.

The current driver implement global Systick interrupt callback
'sys_clock_isr' as usual C function with an argument, while the direct
ARM's interrupt handlers HAVE NOT arguments, they are void functions.

Also direct handler should have some special missing header/footer and
additional attribute.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
cortex_m_systick: remove TICKLESS def, replace CYC_PER_TICK by inline

The TICKLESS define is superfluous and useless, so remove it.

Also replace CYC_PER_TICK based macros by inline const function for the
better type checking, and for the following 'external' clock support.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
cortex_m_systick: use system_write32/system_read32 with DTS defined regs

Switch from C-struct to DTS based register usage due to unpretictable
results with some optimisation. As an example, SYST_CSR (ex CTRL)
register is actually a compound control/status register, so using a
"read-update-write" struct based operations may lead to accidentally
resetting the underrun bit (which is reset by a read operation).

Besides add missing spinlock in the ISR (and refactoring
sys_clock_set_timeout a little bit).

Also update registers names according to the latest Cortex-M ARM
documentations.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
cortex_m_systick: add a new generic file 'cortex-m-systick.yaml'

Move all common properties from dedicated Cortex-M SysTick
related DTS files and put them into the new generic
'cortex-m-systick.yaml' one.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
cortex_m_systick : add "external" clock support

The current driver lacks of external clock support at all, so this
commit fix it.

Fixes #56117

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 commented:
Hi @andyross ,

Sorry to bother you after such a long delay, but I was finally able to get back to the PR and simplify/fix it, so could you review or reassign it to someone else?

Thank you,
Andrey

npitre commented:
I still wonder about this:

```
        min_delay_rvr = (uint64_t)MIN_IRQ_DELAY * SYS_CLOCK_FRQ /
                                sys_clock_hw_cycles_per_sec();
```

Again, `SYS_CLOCK_FRQ` and `sys_clock_hw_cycles_per_sec()` should produce 
the same value.

Maybe you should guard against any discrepancy like this:

```
 #if USE_EXT_CLOCK
 
 #define SYS_CLOCK_FRQ   DT_PROP(DT_NODELABEL(systick), external_frequency)
+BUILD_ASSERT(!IS_ENABLED(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME) &&
+             SYS_CLOCK_FRQ == CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
```

avolkov-1221 commented:
> I still wonder about this:
> 
> ```
>         min_delay_rvr = (uint64_t)MIN_IRQ_DELAY * SYS_CLOCK_FRQ /
>                                 sys_clock_hw_cycles_per_sec();
> ```
> 
> Again, `SYS_CLOCK_FRQ` and `sys_clock_hw_cycles_per_sec()` should produce the same value.

OMG, again copy/paste/aftermerge. Thank you.

> 
> Maybe you should guard against any discrepancy like this:
> 
> ```
>  #if USE_EXT_CLOCK
>  
>  #define SYS_CLOCK_FRQ   DT_PROP(DT_NODELABEL(systick), external_frequency)
> +BUILD_ASSERT(!IS_ENABLED(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME) &&
> +             SYS_CLOCK_FRQ == CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
> ```

Yes, BUILD_ASSERT seems reasonable, as unfortunately SYS_CLOCK_HW_CYCLES_PER_SEC is a board specific and I don't see any way to force them to keep in sync :(.
